Optimize async code in model binders

rynowak opened the pull request:
Optimize the 'no-op' path for our model binders to return a cached task
where possible without going async.

rynowak commented:
OK, now this is really where the sausage gets made (delicious perf-sausage). We can avoid going async in all of the cases where a model binder doesn't actually bind anything, and this helps us avoid allocating a LOT of tasks.

Model binding still isn't in a great place after this change, but it's a big improvement. I nuked `BindingSourceModelBinder` because it's an anti-pattern with this change. You **need** to care about perf if you're writing a model binder unfortunately.

Before (includes changes in #3104):
![image](https://cloud.githubusercontent.com/assets/1430011/9724153/a57bc460-5587-11e5-96da-78ea402e0e4c.png)

After:
![image](https://cloud.githubusercontent.com/assets/1430011/9724159/b1a18928-5587-11e5-80ff-c4ccfb1abfa0.png)


**YOWZA**
This change cuts the number of `Task<ModelBindingResult>` allocations in half. Moreover we avoid allocating a bunch of async state machines we used to have to instantiate - see `BindingSourceModelBinder+<horriblename>` in the first set of results. This change is saving us 8-10mb over 3000 requests.

rynowak commented:
/cc @benaadams eat your :heart: out

benaadams commented:
Now you are cooking with gas! :+1:

dougbu commented on src/Microsoft.AspNet.Mvc.Core/ModelBinding/GenericModelBinder.cs:
```diff
@@ -14,9 +14,20 @@ namespace Microsoft.AspNet.Mvc.ModelBinding
+        private async Task<ModelBindingResult> BindModelCoreAsync(ModelBindingContext bindingContext, Type binderType)
+        {
+            binderType = ResolveBinderType(bindingContext);
             if (binderType != null)
```
Why not pass in `binderType` from `BindModelAsync()`? And why does the value need to be `null`-checked again?

rynowak replied:
It might have decided to become null ¯\_(ツ)_/¯

dougbu commented on src/Microsoft.AspNet.Mvc.Core/ModelBinding/FormCollectionModelBinder.cs:
```diff
@@ -18,7 +18,17 @@ namespace Microsoft.AspNet.Mvc.ModelBinding
+
+        private async Task<ModelBindingResult> BindModelCoreAsync(ModelBindingContext bindingContext)
         {
             if (bindingContext.ModelType != typeof(IFormCollection))
```
Why does this block still exist?

rynowak replied:
Cleaned up

dougbu commented on src/Microsoft.AspNet.Mvc.Core/ModelBinding/BodyModelBinder.cs:
```diff
@@ -14,19 +14,31 @@ namespace Microsoft.AspNet.Mvc.ModelBinding
+        /// <returns>
+        /// A <see cref="Task{ModelBindingResult}"/> which when completed returns a <see cref="ModelBindingResult"/>.
+        /// </returns>
+        protected async Task<ModelBindingResult> BindModelCoreAsync([NotNull] ModelBindingContext bindingContext)
```
`private` ?

rynowak replied:
sure

dougbu commented:
Looks as if `TestBindingSourceModelBinder` still exists and that inherits from `BindingSourceModelBinder`. That adds up to 128 functional test failures.

Does our use of `BindingSourceModelBinder` as a basis for a binder in a functional test imply the `abstract` class would be useful to our customers? Or is `TestBindingSourceModelBinder` another anti-pattern -- a binder that should be checking the `BindingSource` match for itself?

dougbu commented:
:watch:

davidfowl commented:
CI hates this change

rynowak commented:
> Does our use of BindingSourceModelBinder as a basis for a binder in a functional test imply the abstract class would be useful to our customers? Or is TestBindingSourceModelBinder another anti-pattern -- a binder that should be checking the BindingSource match for itself?

I don't think it's worth keeping. We should nuke it from our functional tests.

rynowak commented:
> CI hates this change

Good thing I haven't checked it in yet then. Making all the functional tests happy is usually something I do after addressing PR feedback.

rynowak commented:
Updated

dougbu commented:
:shipit:

davidfowl commented:
@rynowak This change went in right?

rynowak commented:
Yep. So many more perfs now.
